precompiles/xcm/src/lib.rs

```diff
@@ -365,16 +433,567 @@ where
+            .collect::<Vec<MultiAsset>>()
+            .into();
+
+        // Build call with origin.
+        let origin = Some(Runtime::AddressMapping::into_account_id(
+            handle.context().caller,
+        ))
+        .into();
+
+        log::trace!(target: "xcm-precompile:assets_reserve_transfer", "Processed arguments: assets {:?}, amounts: {:?}, beneficiary: {:?}, destination: {:?}, fee_index: {}", assets, amounts, beneficiary, dest, fee_asset_item);
+
+        let call = pallet_xcm::Call::<Runtime>::reserve_transfer_assets {
+            dest: Box::new(dest.into()),
+            beneficiary: Box::new(beneficiary.into()),
+            assets: Box::new(assets.into()),
+            fee_asset_item,
+        };
+
+        // Dispatch a call.
+        RuntimeHelper::<Runtime>::try_dispatch(handle, origin, call)?;
+
+        Ok(succeed(EvmDataWriter::new().write(true).build()))
+    }
+
+    fn send_xcm(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+        let mut input = handle.read_input()?;
+        input.expect_arguments(3)?;
+
+        // Raw call arguments
+        let dest: MultiLocation = input.read::<MultiLocation>()?;
+        let xcm_call: Vec<u8> = input.read::<BoundedBytes<GetXcmSizeLimit>>()?.into();
+
+        log::trace!(target:"xcm-precompile::send_xcm", "Raw arguments: dest: {:?}, xcm_call: {:?}", dest, xcm_call);
+
+        let xcm = xcm::VersionedXcm::<()>::decode_all_with_depth_limit(
+            xcm::MAX_XCM_DECODE_DEPTH,
+            &mut xcm_call.as_slice(),
+        )
+        .map_err(|_| revert("Failed to decode xcm instructions"))?;
+
+        // Build call with origin.
+        let origin = Some(Runtime::AddressMapping::into_account_id(
+            handle.context().caller,
+        ))
+        .into();
+        let call = pallet_xcm::Call::<Runtime>::send {
+            dest: Box::new(dest.into()),
+            message: Box::new(xcm),
+        };
+        log::trace!(target: "xcm-send_xcm", "Processed arguments:  XCM call: {:?}", call);
+        // Dispatch a call.
+        RuntimeHelper::<Runtime>::try_dispatch(handle, origin, call)?;
+
+        Ok(succeed(EvmDataWriter::new().write(true).build()))
+    }
+
+    fn transfer(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+        let mut input = handle.read_input()?;
+        input.expect_arguments(5)?;
+
+        // Read call arguments
+        let currency_address = input.read::<Address>()?;
+        let amount_of_tokens = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting amount_of_tokens, maybe value too large"))?;
+        let destination = input.read::<MultiLocation>()?;
+        let weight = input.read::<WeightV2>()?;
+
+        let asset_id = Runtime::address_to_asset_id(currency_address.into())
+            .ok_or(revert("Failed to resolve fee asset id from address"))?;
+        let dest_weight_limit = if weight.is_max() {
```

Comment:
The interface docstring should include when `Unlimited` is used.

Personally, I think `zero` is a more conventional value. For max weight, it has indicated unlimited by using itself.

Reply:
I have added the info about Unlimited weight in interface.

> Personally, I think zero is a more conventional value. For max weight, it has indicated unlimited by using itself.

Not sure what you mean by this.

Reply:
> > Personally, I think zero is a more conventional value. For max weight, it has indicated unlimited by using itself.
> 
> Not sure what you mean by this.

I mean weight `max` vs `zero` to indicate `Unlimited`. `is_zero` should be used instead of `is_max`.

Reply:
for this case `is_max` make more sense imo. Can we go ahead with this?

Reply:
This is part of the public API so let's be cautious.

Zero value is conventional to indicate unlimited/unstrict. Using a `uint64` max value is counter-intuition and clumsy to use. In addition, when a max weight is passed, there is actually no need to convert it to `Unlimited` as it's unlimited itself.

Reply:
Agree with Shaun about `zero` as the sentinel value.

It's harder to mess up, even if in the future we change some things related to weight type (probably won't, but still). Zero will always be a zero.

Reply:
Make sense, I was not able to understand what exactly you meant but now I get it.
So you are saying that `Weight = 0` , should result into `WeightLimit::Unlimited` and not `u64::MAX` value.
So what you are suggesting is this

```
let dest_weight_limit = if weight.is_zero() {
            WeightLimit::Unlimited
        } else {
            WeightLimit::Limited(weight.get_weight())
        };
        
 ```
I got the idea of using `u64::MAX` for Unlimited from MB but this is a better approach, I agree.

Reply:
Yes, correct.